cot/src/project.rs

```diff
@@ -87,605 +155,1439 @@ pub trait CotApp: Send + Sync {
-    /// access to the project's configuration, apps, database, etc. An example
-    /// of such middleware is the
-    /// [`StaticFilesMiddleware`](cot::static_files::StaticFilesMiddleware),
-    /// which iterates through all the registered apps and collects the
-    /// static files from them.
+impl ProjectContext<WithApps> {
+    /// Returns the configuration for the project.
     ///
     /// # Examples
     ///
     /// ```
-    /// use cot::static_files::StaticFilesMiddleware;
-    /// use cot::{CotApp, CotProject};
+    /// use cot::request::{Request, RequestExt};
+    /// use cot::response::Response;
     ///
-    /// # #[tokio::main]
-    /// # async fn main() -> cot::Result<()> {
-    /// let cot_project = CotProject::builder()
-    ///     .middleware_with_context(StaticFilesMiddleware::from_app_context)
-    ///     .build()
-    ///     .await?;
+    /// async fn index(request: Request) -> cot::Result<Response> {
+    ///     let config = request.context().config();
+    ///     // can also be accessed via:
+    ///     let config = request.project_config();
     ///
-    /// # Ok(())
-    /// # }
+    ///     let db_url = &config.database.url;
+    ///
+    ///     // ...
+    /// #    todo!()
+    /// }
     /// ```
     #[must_use]
-    pub fn middleware_with_context<M, F>(
-        self,
-        get_middleware: F,
-    ) -> CotProjectBuilder<IntoCotError<IntoCotResponse<M::Service>>>
-    where
-        M: Layer<RouterService>,
-        F: FnOnce(&AppContext) -> M,
-    {
-        self.into_builder_with_service()
-            .middleware_with_context(get_middleware)
+    pub fn config(&self) -> &ProjectConfig {
+        &self.config
     }
 
-    /// Builds the Cot project instance.
+    /// Returns the apps for the project.
     ///
     /// # Examples
     ///
-    /// ```no_run
-    /// use cot::{CotApp, CotProject};
+    /// ```
+    /// use cot::request::{Request, RequestExt};
+    /// use cot::response::Response;
     ///
-    /// #[cot::main]
-    /// async fn main() -> cot::Result<CotProject> {
-    ///     let cot_project = CotProject::builder().build().await?;
+    /// async fn index(request: Request) -> cot::Result<Response> {
+    ///     let apps = request.context().apps();
     ///
-    ///     Ok(cot_project)
+    ///     // ...
+    /// #    todo!()
     /// }
     /// ```
-    pub async fn build(self) -> cot::Result<CotProject> {
-        self.into_builder_with_service().build().await
+    #[must_use]
+    pub fn apps(&self) -> &[Box<dyn App>] {
+        &self.apps
     }
 
     #[must_use]
-    fn into_builder_with_service(mut self) -> CotProjectBuilder<RouterService> {
-        let router = Arc::new(Router::with_urls(self.urls));
-        self.context.router = Arc::clone(&router);
-
-        CotProjectBuilder {
-            context: self.context,
-            cli: self.cli,
-            urls: vec![],
-            handler: RouterService::new(router),
+    fn into_initialized(
```

Comment:
To keep the naming consistent, this should be called `with_backend` or something similar